[WIP] Binary Operators

wmalpica opened the pull request:
Implemented so far is the first implementation concept using nested templating. It has been implemented with support for literals and 14 different binary operators. 

We would like feedback on the API implemented, which is:

``
gdf_error gdf_scalar_operation(gdf_column* out, gdf_column* vax, gdf_scalar* vay, gdf_scalar* def, gdf_binary_operator ope);

gdf_error gdf_vector_operation(gdf_column* out, gdf_column* vax, gdf_column* vay, gdf_scalar* def, gdf_binary_operator ope);

enum gdf_binary_operator {
    GDF_ADD,
    GDF_SUB,
    GDF_MUL,
    GDF_DIV,
    GDF_TRUE_DIV,
    GDF_FLOOR_DIV,
    GDF_MOD,
    GDF_POW,
    //GDF_COMBINE,
    //GDF_COMBINE_FIRST,
    //GDF_ROUND,
    GDF_EQUAL,
    GDF_NOT_EQUAL,
    GDF_LESS,
    GDF_GREATER,
    GDF_LESS_EQUAL,
    GDF_GREATER_EQUAL,
    //GDF_PRODUCT,
    //GDF_DOT
};


struct gdf_scalar {
    union gdf_data {
        std::int8_t  si08;
        std::int16_t si16;
        std::int32_t si32;
        std::int64_t si64;
        float        fp32;
        double       fp64;
    };

    gdf_data  data;
    gdf_dtype type;
};
``

We have already started on a second implementation using NVRTC which uses the same API.

Unit tests have not yet been implemented and will be implemented after an implementation method has been decided.

This current implementation right now only supports the second operand being able to be a scalar, which can be improved upon easily later.

In order to compile the binary operation functionality, use the cmake variable "BINARY_OPERATION_VERSION" and choose the binary version.

Example:
cmake -DCMAKE_BUILD_TYPE=Release -DBINARY_OPERATION_VERSION:STRING=V1 ../../code/libgdf

This cmake switch was added right now because compilation time is very large due to all the functions generated by the templated code. For example compilation on a laptop with an 8-core i7 processor and 16Gb ram was:
    Time to compile: 77m10.996s
    Ram memory used: More than 13.5GB
    Release binary size: 69,931,016 bytes

Commit:
Binary Operators
Implemented first approach, multiple dispatch and generic programming.

scopatz commented:
This has a conflict with master now

Commit:
Binary Operations
The 2nd approach using NVRTC ("Jitify" library).
Base version for review.

Commit:
Binary Operations
The 2nd approach using NVRTC ("Jitify" library).
Base and stable version of the 2nd approach.

Commit:
Binary Operations
The 2nd approach using NVRTC ("Jitify" library).
Added kernels for binary operations with default values.

Commit:
Binary Operations
The 2nd approach using NVRTC ("Jitify" library).
Improved kernel implementation. Scalar type will be converted at kernel execution (type defined).

Commit:
Binary Operations
The 2nd approach using NVRTC ("Jitify" library).
Added all base types to gdf_dtypes.

Commit:
Binary Operations
The 2nd approach using NVRTC ("Jitify" library).
Changed gdf_scalar interface.

Commit:
Binary Operations
The 2nd approach using NVRTC ("Jitify" library).
Erased 'valid' field input for kernels without a default value.

Commit:
Binary Operations
The 2nd approach using NVRTC ("Jitify" library).
Updated valid field in kernels and added integration tests for all kernels.

Commit:
Binary Operations
Using Jitify utility to choose the optimal block size at kernel calls.

Commit:
Binary Operations
Fixed gtest compilation issues.

Commit:
Binary Operations
Fixed gtest issues

Commit:
Binary Operations
Added shared libraries for tests.

Commit:
Binary Operations
The 2nd approach using NVRTC ("Jitify" library).
Erased unused functionality.

Commit:
Binary Operations
The 2nd approach using NVRTC ("Jitify" library).
Added binary operations.

harrism commented on src/binary/binary2/kernel.cpp:
```diff
@@ -0,0 +1,141 @@
+
+    template <typename TypeOut, typename TypeVax, typename TypeVay, typename TypeOpe>
+    __global__
+    void kernel_v_s(int size, TypeOut* out_data, TypeVax* vax_data, gdf_data vay_data) {
```
Shouldn't `pdf_data` on this line be `TypeVay`?

ironbit replied:
The operands types are column and scalar. gdf_data (scalar - union) is passed to the kernel. Inside the kernel, it is used a type cast operator when the operation is performed.

harrism commented on include/gdf/cffi/types.h:
```diff
@@ -16,6 +20,23 @@ typedef enum {
     N_GDF_TYPES, /* additional types should go BEFORE N_GDF_TYPES */
 } gdf_dtype;
 
+union gdf_data {
```
Are the names in this union required to only be four characters for some reason? It would be nice to spell them out better, especially "invd", "tmst", "dt32" and "dt64".

ironbit replied:
There isn't a particular reason. Is there any name style and code style for this project?

harrism replied:
Human readable variable names is good practice for every project.

harrism commented on src/binary/binary2/kernel.cpp:
```diff
@@ -0,0 +1,141 @@
+        }
+    }
+
+    template <typename TypeOut, typename TypeVax, typename TypeVay, typename TypeDef, typename TypeOpe>
```
Wow, `TypeDef` is very risky typename, considering its proximity to `typedef`. Could just use TDef, TVax, etc. Or reorder, putting `Type` last in the names.

ironbit replied:
No problem, it will be changed.

harrism commented on src/binary/binary2/kernel.cpp:
```diff
@@ -0,0 +1,141 @@
+
+    template <typename TypeOut, typename TypeVax, typename TypeVay, typename TypeDef, typename TypeOpe>
+    __global__
+    void kernel_v_s_d(int size, gdf_data def_data,
```
Should `gdf_data` on this line be `TypeDef`?

harrism commented on src/binary/binary2/kernel.cpp:
```diff
@@ -0,0 +1,141 @@
+            AbstractOperation<TypeOpe> operation;
+            out_data[i] = operation.template operate<TypeOut, TypeVax, TypeVay>(vax_data_aux, (TypeVay)vay_data);
+
+            __syncwarp();
```
What is the reason for `__syncwarp()` here? I don't see any sharing of data between threads. Is it necessary? Same question for all the other `__syncwarp()` calls.

ironbit replied:
It's not correct. It'll be removed.

harrism commented on src/binary/binary2/launcher.cpp:
```diff
@@ -0,0 +1,115 @@
+        return *this;
+    }
+
+    Launcher& Launcher::instantiate(gdf_column* out, gdf_column* vax, gdf_column* vay, gdf_binary_operator ope) {
```
Since each of these `instantiate` functions varies only in the type of one argument, and their bodies are identical, couldn't you template them to reduce redundancy? You might even use variadic templates to collapse all into one instantiate and one launch function.

ironbit replied:
That's a good point

harrism commented on src/binary/binary2/operation.cpp:
```diff
@@ -0,0 +1,196 @@
+            return static_cast<ConcreteOperation*>(this)->template operate<TypeOut, TypeVax, TypeVay>(x, y);
+        }
+    };
+/*
```
Do you need to keep the commented-out code?

ironbit replied:
I implemented this version as a "prototype". It requires code style, name style, code organization, better testing, etc. Whether a requirement could be underflow detection, the commented code would be useful.

harrism commented on src/binary/binary2/traits.cpp:
```diff
@@ -0,0 +1,216 @@
+R"***(
+#pragma once
+
+    struct IntegralSigned {};
```
Hmmmm, I believe Jitify already has some support for std::type_traits, e.g. https://github.com/NVIDIA/jitify/blob/master/jitify.hpp#L1124
Do you need to roll your own?

ironbit replied:
I had some issue with that. I expressed it in my daily scrum. It is supposed to be working, but it isn't. I included the libraries, however "common_type" and others aren't working, because of the error messages using nvrtc.

harrism replied:
Let's work with Kate Clark and Ben Barsdell on making Jitify better. Are you in touch with them?

ironbit replied:
No. I'm not. I remember having issues related to "type_traits" header and with compiler option "--device-as-default-execution-space", which wasn't working.

harrism replied:
I will contact them.

harrism commented on src/tests/binary-operation/library/operation.h:
```diff
@@ -0,0 +1,91 @@
+            return (TypeOut)pow((double)vax, (double)vay);
+        }
+    };
+}
```
It's nice in this situation (closing brackets far from their openings) to comment what is being closed, e.g.:
```
} // namespace operation
} // namespace test
} // namespace gdf
```

ironbit replied:
Documentation is important. I was focused on implementation.

harrism reviewed (changes requested):
My biggest concern with this PR is that there is a lot of code and basically zero documentation. First, I think it needs high-level documentation of the functionality and organization, and it needs operational documentation for the main APIs being introduced (both external and internal APIs).

ironbit committed:
Merge branch 'master' into binary-operators-draft

Commit:
Binary Operations
The 2nd approach using NVRTC ("Jitify" library).
Erased syncwarp primitive.

Commit:
Merge branch 'binary-operators-draft' of github.com:BlazingDB/libgdf into binary-operators-draft

Commit:
Binary Operations
Changed submodule "Jitify" to https protocol.

Commit:
Binary Operations
Added CUDA configuration for NVRTC library.

Commit:
Binary Operations
Added 'stubs' directory in cmake for CUDA library.

Commit:
Binary Operations
Changed default string variable for selection of binary operations.

Commit:
Binary Operations
Added google benchmark for binary operations - NVRTC implementation.

Commit:
Binary Operations
Changed interface for cffi python.

Commit:
Binary Operations
Solved issues related to other tests due to a fix in the storage duration of Jitify cache.

Commit:
Binary Operations
Added 'stubs' directory in the library path (python) for NVRTC and CUDA libraries.

Commit:
Binary Operations
Improved vector library code.

scopatz commented on include/gdf/cffi/types.h:
```diff
@@ -91,6 +116,29 @@ typedef enum {
+    GDF_POW,
+    //GDF_COMBINE,
+    //GDF_COMBINE_FIRST,
+    //GDF_ROUND,
```
I don't think that it makes sense to have intermediate parts of an enum commented out.

scopatz replied:
This can lead to API instabilities.

ironbit replied:
I'm so sorry, I hadn't seen your message.
It is used only as an enum (not casting to integers).
It could be error-prone. In any case, it will be moved to the end.

Commit:
Binary Operations
Improved testing for binary operations.

Commit:
Added libgdf static library target as optional (disabled by default).

Commit:
Binary Operations
Added comments to benchmark.

Commit:
Changed the name of the static library (libgdf.a).

aucahuasi committed:
Minor fixes for LIBGDF_STATIC_LIB and update README.md

aucahuasi committed:
Link gdf against cuda and nvrtc libs

Commit:
Binary Operations
Fixed issue in testing libraries.

aucahuasi committed:
Add cuda runtime libs

aucahuasi committed:
Merge branch 'binary-operators-draft' of https://github.com/BlazingDB/libgdf into binary-operators-draft

Commit:
Merge branch 'master' into binary-operators-draft

Commit:
Updated branch 'binary-operators-draft'

Commit:
Binary Operations
Added position independent code (PIC) for shared library (libgdf).

Commit:
Binary Operations
Erased binary operation version one due to an incompatibility with the new interface.

Commit:
Binary Operations
Updated the README.md file with the benchmark information.

Commit:
Binary Operations
Added license & copyright for BlazingDB.

Commit:
Binary Operations
Added the cuda stubs folder in the environment variable LD_LIBRARY_PATH for Travis.

Commit:
Binary Operations
Improved some parts of the code.

aucahuasi committed:
test solution

aucahuasi committed:
upgrade py package

nsakharnykh commented:
what is the corresponding pygdf PR for the bindings update?

Commit:
Binary Operations
Added verification of the inputs for the binary operations.

Commit:
Merge branch 'master' into binary-operators-draft

Commit:
Merge branch 'master' into binary-operators-draft

Commit:
Merge branch 'binary-operators-draft' of github.com:BlazingDB/libgdf into binary-operators-draft

Commit:
Binary Operations
Erased invalid field in union gdf_data.

Commit:
Merge branch 'master' into binary-operators-draft

Commit:
Binary Operations
Fixed issue related to inputs verification.

nsakharnykh commented on cmake/Templates/GoogleBenchmark.CMakeLists.txt.cmake:
```diff
@@ -0,0 +1,36 @@
+#=============================================================================
```
Can we use this for CI? @mike-wendt

nsakharnykh commented on include/gdf/cffi/types.h:
```diff
@@ -120,6 +159,36 @@ typedef enum {
+    GDF_GREATER,
+    GDF_LESS_EQUAL,
+    GDF_GREATER_EQUAL,
+    //GDF_COMBINE,
```
Add a description/comment why those are not implemented

scopatz replied:
Or, preferably, please remove.

nsakharnykh commented on src/binary/CMakeLists.txt:
```diff
@@ -0,0 +1,28 @@
+message(STATUS "Binary Operation Version: V2 - NVRTC")
+
+set(source_files
+    src/binary/binary2/binary.cpp
```
Rename `binary2` folder to `binary_nvrtc` or `binary_jitify`

ironbit replied:
The source code has been reorganized, that's why I think it's convenient to reply to every observation.
https://github.com/BlazingDB/libgdf/tree/binary-operators-draft/src/binary-operation/jit

nsakharnykh commented on src/binary/binary2/kernel.cpp:
```diff
@@ -0,0 +1,155 @@
+
+    template <typename TypeOut, typename TypeVax, typename TypeVay, typename TypeOpe>
+    __global__
+    void kernel_v_s(int size, TypeOut* out_data, TypeVax* vax_data, gdf_data vay_data) {
```
We need `kernel_s_v` as well - should be 6 combinations overall, one kernel per the top level GDF function

ironbit replied:
Implemented the remaining scalar-vector operations.

nsakharnykh commented on src/binary/binary2/launcher.cpp:
```diff
@@ -0,0 +1,94 @@
+#include "binary/binary2/cuda.h"
+
+namespace gdf {
+    static thread_local jitify::JitCache JitCache;
```
We should be able to store JitCache when exiting the process to store in a file, then use an env variable to read it from the specified file to have the cache enabled between different processes

nsakharnykh commented on src/binary/binary2/operation.cpp:
```diff
@@ -0,0 +1,213 @@
+        }
+    };
+/*
+    struct Add : public AbstractOperation<Add> {
```
Should add a comment with the description why this could be useful, i.e. for the overflows - if we ever wanted to enable this "feature" in the future

ironbit replied:
The source code has been optimized for jit.
The commented code has been moved outside the jit code and it has been added a description.
https://github.com/BlazingDB/libgdf/blob/96012d91aff817e6105a8fa8983e1587d237493f/src/binary-operation/jit/code/operation.cpp#L159

nsakharnykh commented on src/library/field.h:
```diff
@@ -0,0 +1,109 @@
+/*
```
Move `library` under `bench/binary` or `tests/binary` since it's only used in tests, probably also rename to something like binary_helper

ironbit replied:
https://github.com/BlazingDB/libgdf/tree/binary-operators-draft/src/tests/binary-operation/util

nsakharnykh commented on src/bench/binary-operation-benchmark.cpp:
```diff
@@ -0,0 +1,203 @@
+/*
```
Needs to be moved to `src/bench/binary` to have a folder per libgdf operator (one folder for binary, one for parquet, one for joins, etc.)

ironbit replied:
https://github.com/BlazingDB/libgdf/tree/binary-operators-draft/src/bench/binary-operation

nsakharnykh commented on src/binary/binary2/kernel.cpp:
```diff
@@ -0,0 +1,155 @@
+
+        for (int i=start; i<size; i+=step) {
+            AbstractOperation<TypeOpe> operation;
+            out_data[i] = operation.template operate<TypeOut, TypeVax, TypeVay>(vax_data[i], (TypeVay)vay_data);
```
We need to set the output valid bit mask - I don't see it being handled in the code. It should be an OR between the two bit masks of the two input operands.

ironbit replied:
The output valid bitmask is processed in all kernels.
The kernels also have been optimized using 'Bit Hacks'. In the benchmark, the kernel (v_v_v_d) reduces its time approx. in 5us in all of its benchmarks, while the kernel (v_v_v) increments its time approx. in 7us due to the bitmask processing. 
https://github.com/BlazingDB/libgdf/blob/729cbfac6ae2281894b99644c58643184bd18d85/src/binary-operation/jit/code/kernel.cpp#L82

nsakharnykh reviewed (changes requested):
We did a live PR review with @jrhemstad @williamBlazing and @ironbit - requesting more changes. Also, we need the python bindings updated to use the new binary ops API: @mt-jones, can you post an update on where we are with this?

Commit:
Binary Operations
Reorganized the source code.

Commit:
Binary Operations
Optimized the source code in Jit.

Commit:
Binary Operations
Added namespace documentation.

scopatz commented on src/binary-operation/jit/util/type.cpp:
```diff
@@ -0,0 +1,96 @@
+                return "Pow";
+            //case GDF_COMBINE:
+            //case GDF_COMBINE_FIRST:
+            //case GDF_ROUND:
```
Are these simply not implemented? Probably should remove or make a comment

Commit:
Binary Operations
Moved 'util' library used in tests and benchmark.

Commit:
Merge branch 'master' into binary-operators-draft

Commit:
Binary Operations
Optimized Jit kernel code (bit hacks) and added the output valid bitmask in some kernels.

Commit:
Merge branch 'master' into binary-operators-draft

Commit:
Binary Operations
Added documentation related to the new interface.
Changed the interface ('valid' field in gdf_scalar).

Commit:
Merge branch 'master' into binary-operators-draft

Commit:
Merge branch 'master' into binary-operators-draft

Commit:
Binary Operations
Added the changes for scalar vector operations.

Commit:
Binary Operations
Added the standard header for math functions.

Commit:
Binary Operations
Added a 'libcuda' soft link.

Commit:
new order by api

aocsa committed:
[binary-operators-draft]: udpate get_column_byte_width to support unsigned types

aocsa committed:
[binary-operator-draft] update gpu_apply_stencil to work with least-significant bit  format

Commit:
fixed some issues with binary ops not being able to output if the space was too large

Commit:
made a small fix to the sorting code so it gets the size properly, it ws working before but could have had weird errors at sizes above 1bn records

Commit:
removed some couts

Commit:
removed some couts

aucahuasi committed:
TODO use a fixed compute capability
